### DISCUSSION TITLE
MFRONT - How To create fourth order tensor from four vectors

### DISCUSSION THREAD

**Discussion by kkolari** [thelfer/tfel#434] [General]
How to create fourth order tensor in MFront:
C_ijkl = a_i * b_j * c_k * d_l
I've tried to find syntax/examples but with no success
thanks for advance!

**Comment by thelfer** [comment]
@kkolari Well it is very complicated. 

- first, fourth order tensors are stored as matrices.
- secund, there are four kinds of fourth order tensors (st2tost2, t2tot2, t2tost2, st2tot2)
- third, their number of components depends on the modelling hypothesis

For example, an st2tost2 tensor is a linear transformation of a symmetric order tensor in a symmetric order tensor. It has 9 components in 1D, 36 components in 3D.

We thus generally don't use indices to define a fourth order tensor.

**Reply by thelfer** [reply]
@kkolari to be definitively complete, there are many ways to manipulate fourth order tensors in `TFEL/Math`. We just focuse on the tensorial operations to define them rather than manipulating indices. The good question is thus: what do you want to do ?

**Reply by kkolari** [reply]
Thanks to @thelfer for prompt reply - again. Is there a web page with examples about tensor formulation/manipulation/multiplication/etc.  I am not very well familiar with the MFront syntax.
For example, I want to create C_ijkl ( I am using indexes to get the idea what I am looking for)
C_ijkl = N_i N_j K_kl
where K_kl is symmetric, positive definite second order tensor (3x3)
and N_k is a vector of order 3

**Reply by thelfer** [reply]
> Is there a web page with examples about tensor

https://thelfer.github.io/tfel/web/tensors.html

> I want to create C_ijkl ..

How will you use this ?

you can create a symmetric tensor from vector N using the `stensor::buildFromVectorDiadicProduct`.

```c++
const auto NxN = stensor<N, real>::buildFromVectorDiadicProduct(N);
```

then

```c++
const auto C = (NxN) ^ (K);
```

**Reply by kkolari** [reply]
Thank you. Exactly what I needed..
Using it in the formulation of Secant Stiffness (with anisotropic damaging into direction of N_k)
